collectors/perf.cpp

```diff
@@ -443,88 +465,25 @@ bool PerfCollector::collect(int64_t now)
     return true;
 }
 
-bool PerfCollector::perf_counter_pause() {
-#if defined(__aarch64__)
-    asm volatile("mrs %0, PMCNTENSET_EL0" : "=r" (PMCNTENSET_EL0_safe));
-    // stop counters for arm64
-    asm volatile("mrs %0, PMCR_EL0" : "=r" (PMCR_EL0_safe));
-    asm volatile("msr PMCR_EL0, %0" : : "r" (PMCR_EL0_safe & 0xFFFFFFFFFFFFFFFE));
-#elif defined(__arm__)
-    asm volatile("mrc p15, 0, %0, c9, c12, 1" : "=r"(PMCNTENSET_EL0_safe));
-    // stop counters for arm32
-    asm volatile("mrc p15, 0, %0, c9, c12, 0" : "=r"(PMCR_EL0_safe));
-    asm volatile("mcr p15, 0, %0, c9, c12, 0" : : "r"(PMCR_EL0_safe & 0xFFFFFFFE));
-#endif
-    return true;
-}
-
-bool PerfCollector::perf_counter_resume() {
-#if defined(__aarch64__)
-    // start counters for arm64
-    asm volatile("msr PMCNTENSET_EL0, %0" : : "r" (PMCNTENSET_EL0_safe));
-    asm volatile("msr PMCR_EL0, %0" : : "r" (PMCR_EL0_safe));
-#elif defined(__arm__)
-    // start counters for arm32
-    asm volatile("mcr p15, 0, %0, c9, c12, 1" : : "r"(PMCNTENSET_EL0_safe));
-    asm volatile("mcr p15, 0, %0, c9, c12, 0" : : "r"(PMCR_EL0_safe));
-#endif
-    return true;
-}
-
-
-bool PerfCollector::collect_scope_start(int64_t now, uint16_t func_id, int32_t flags) {
+bool PerfCollector::collect_scope_start(uint16_t func_id, int32_t flags, int tid) {
 #if defined(__x86_64__)
     if (!attempt_collect_scope_x64) {
         attempt_collect_scope_x64 = true;
         DBG_LOG("WARNING: Frequent invocation of collect_scope on x64 devices may introduce "
                 "significant overhead to the kernel perf counter data.\n");
     }
 #endif
-    if (!perf_counter_pause()) return false;
     if (!mCollecting) return false;
     struct snapshot snap;
     if (flags & COLLECT_REPLAY_THREADS || flags & COLLECT_ALL_THREADS)
     {
-        for (perf_thread &t : mReplayThreads)
-        {
-            t.eventCtx.collect_scope(now, func_id, false);
-        }
-    }
-    if (flags & COLLECT_BG_THREADS || flags & COLLECT_ALL_THREADS)
-    {
-        for (perf_thread &t : mBgThreads)
-        {
-            t.eventCtx.collect_scope(now, func_id, false);
-        }
-    }
-    if (flags & COLLECT_MULTI_PMU_THREADS || flags & COLLECT_ALL_THREADS)
-    {
-        for (perf_thread &t : mMultiPMUThreads)
-        {
-            t.eventCtx.collect_scope(now, func_id, false);
-        }
-    }
-    if (flags & COLLECT_BOOKER_THREADS || flags & COLLECT_ALL_THREADS)
-    {
-        for (perf_thread &t : mBookerThread)
-        {
-            t.eventCtx.collect_scope(now, func_id, false);
-        }
-    }
-    if (flags & COLLECT_CSPMU_THREADS || flags & COLLECT_ALL_THREADS)
-    {
-        for (perf_thread &t : mCSPMUThreads)
-        {
-            t.eventCtx.collect_scope(now, func_id, false);
-        }
+        mReplayThreads[tid].eventCtx.collect_scope(func_id, false, get_pmu_bits());
```

Comment:
1. consider multiple pmus
2. mReplayThreads is vector, tid is not the index of mReplayThreads.

Reply:
Thanks for your advice. But the per api new function only want to focus on the CPU cycles. And it will only read cycles event counter register.  So I wonder if it is not necessary to support multi-events in this case?